NEXT-15111 - Added concept parts for data abstraction layer (#267)
shopware/docs: guides/plugins/plugins/framework/store-api/add-caching-for-store-api-route.md

# Add caching for store api route

## Overview

> In this guide you'll learn how to add a cache layer to your custom store API route. In this example, we will add a cache layer for the `ExampleRoute`, which created in the [Add store api route](./add-store-api-route.md) guide. For the cache invalidation we will write a invalidation subscriber.

Reviewer: ```suggestion
In this guide you'll learn how to add a cache layer to your custom store API route. In this example, we will add a cache layer for the `ExampleRoute`, which is created in the [Add store api route](./add-store-api-route.md) guide. For the cache invalidation we will write a invalidation subscriber.
```

Reviewer 2: "Store API"

## Prerequisites

In order to add a cache layer for the store api route, you first need a store api route as base. Therefore, you can refer to the [Add store api route](./add-store-api-route.md) guide.

You also should have a look at our [Adding custom complex data](../data-handling/add-custom-complex-data.md) guide, since this guide is built upon it.

## Add cache layer

> As you may already learned from the [Add store api route](./add-store-api-route.md) guide, we use abstract classes to make our routes more decoratable.

Reviewer: ```suggestion
As you might have learned already from the [Add store api route](./add-store-api-route.md) guide, we use abstract classes to make our routes more decoratable.
```

This concept is very advantageous if we now want to include a cache layer for the route. There are of course different ways to do this - but in this guide we show how we implemented it in the core.

### Add cached route class
First, we create an abstract class called `CachedExampleRoute` which extends the `AbstractExampleRoute`.

> {% code title="<plugin root>/src/Core/Content/Example/SalesChannel/CachedExampleRoute.php" %}

Reviewer: Can you add the respective services.xml entry?
You can do so by using tabs:

```
{% tabs %}
{% tab title="CachedExampleRoute.php" %}
{% code title="<plugin root>/src/Core/Content/Example/SalesChannel/CachedExampleRoute.php" %}
// PHP Code here
{% endcode %}
{% endtab %}

{% tab title="services.xml" %}
{% code title="<plugin root>/src/Resources/config/services.xml" %}
// services.xml here
{% endcode %}
{% endtab %}
{% endtabs %}
```

For the services.xml, just the single entry will suffice:
```
<?xml version="1.0" ?>
<container xmlns="http://symfony.com/schema/dic/services"
           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">

    <services>
        <service id="Swag\BasicExample\Core\Content\Example\SalesChannel\CachedExampleRoute">
            ...
        </service>
    </services>
</container>
```

        ));

        $this->cache->save($item);

        return $response;
    }
    
    public static function buildName(): string 
    {
        return 'example-route';
    }
  
    private function generateKey(SalesChannelContext $context, Criteria $criteria): string
    {
        $parts = [
            self::buildName(),
            // generate a hash for the route criteria
            $this->generator->getCriteriaHash($criteria),
            // generate a hash for the current context 
            $this->generator->getSalesChannelContextHash($context),
        ];
          
        return md5(JsonFieldSerializer::encodeJson($parts));
    }
}
```
{% endcode %}

In the new `CachedExampleRoute` some core classes are used which simplify the caching.
* `TagAwareAdapterInterface` - Used to read, write and tag cache items.
* `EntityCacheKeyGenerator` - Used to generate hashes for the context and/or criteria;
* `AbstractCacheTracer` - Traces all system config keys that were accessed. The data is needed later for cache invalidation.
* `CacheCompressor` - Provides an optimal compression of the cache entries to use as little disk space as possible.

### Add cache invalidation

> Cache invalidation is many times harder to implement than the actual caching. Finding the right balance between too much and too little invalidation is difficult.

Reviewer: ```suggestion
Cache invalidation is much harder to implement than the actual caching. Finding the right balance between too much and too little invalidation is difficult.
```

Therefore, there is no precise guidance or documentation on when to invalidate what. What and how to invalidate depends on what has been cached. For example, the product routes in the core are always invalidated when the product is written, but also when the product is ordered and reaches the out-of-stock status.
The entire cache invalidation in Shopware is controlled via events. On the one hand there is the entity written event and on the other hand the corresponding business events like `ProductNoLongerAvailableEvent`.

> {% code title="<plugin root>/src/Core/Content/Example/SalesChannel/CacheInvalidationSubscriber.php" %}

Reviewer: Same here with the services.xml and the tabs
